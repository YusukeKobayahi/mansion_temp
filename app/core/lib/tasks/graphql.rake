# frozen_string_literal: true

module GraphqlTask
  extend Rake::DSL

  namespace :graphql do
    namespace :consumer do
      task schema_dump: :environment do
        require 'graphql/rake_task'

        GraphQL::RakeTask.new(
          load_schema: lambda { |_task|
            require File.expand_path('../../app/graphql/consumer/inno_schema', __dir__)
            Consumer::InnoSchema
          },
          directory: './graphql/consumer',
        )
        Rake::Task['graphql:schema:idl'].invoke
      end
    end
  end
end
