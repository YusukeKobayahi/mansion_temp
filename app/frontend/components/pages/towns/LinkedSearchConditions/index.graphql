fragment MsLibraryTownsLinkedSearchConditions on Town {
  linkedSearchConditions {
    key
    value
    name
  }
}
