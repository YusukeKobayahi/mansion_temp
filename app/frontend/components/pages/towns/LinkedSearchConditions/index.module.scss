@import "~/styles/variables";
@import "~/styles/utils";

.container {
  margin-top: 20px;
  border: 1px solid #e5e5e5;
  border-radius: 4px;
}

.title {
  border-radius: 4px 4px 0 0;
  background-color: #f4f4f4;
  padding: 0 10px;
  position: relative;
  z-index: 10;

  p {
    line-height: 40px;
    font-size: 16px;
    color: #373a3c;
  }
}

.list {
  padding: 20px;
  display: flex;
  flex-wrap: wrap;

  .btn {
    padding: 0;
    background-color: unset;
    color: #b2a173;
    text-decoration: underline;
    margin-right: 20px;
  }
}
