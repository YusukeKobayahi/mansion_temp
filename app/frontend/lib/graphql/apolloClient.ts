// 参考
// https://github.com/vercel/next.js/tree/canary/examples/with-apollo

import { useMemo } from "react";
import { ApolloClient, ApolloLink, InMemoryCache } from "@apollo/client";
import { createHttpLink } from "apollo-link-http";
import { from } from "apollo-link";
import { createFragmentArgumentLink } from "apollo-link-fragment-argument";
import { mergeDeep, relayStylePagination } from "@apollo/client/utilities";
import { isSSR } from "~/lib/utils";

/* eslint-disable @typescript-eslint/no-explicit-any -- キャッシュの型を事前に決められないため */
type CacheState = Record<string, any> | null;
/* eslint-enable @typescript-eslint/no-explicit-any */

/* eslint-disable @typescript-eslint/naming-convention -- GraphQLのタイプ名 (このconventionに従わない)を指定したいので */
const createCache = () =>
  new InMemoryCache({
    typePolicies: {
      Mansion: {
        keyFields: ["uniqueCode"],
        fields: {
          salesHistories: relayStylePagination(),
        },
      },
    },
  });
/* eslint-enable @typescript-eslint/naming-convention */

const link = from([
  createFragmentArgumentLink(),
  createHttpLink({
    uri: `${process.env.NEXT_PUBLIC_CORE_API_URL}/consumer/graphql`,
    fetchOptions: {
      mode: "cors",
    },
  }),
]);

function createApolloClient() {
  return new ApolloClient({
    ssrMode: isSSR(),
    // See: https://github.com/apollographql/apollo-link/issues/1258#issuecomment-604031488
    // TODO: @apollo/clientバージョン上げたら直るかも
    link: (link as unknown) as ApolloLink,
    cache: createCache(),
    defaultOptions: {
      watchQuery: {
        // SSR時にuseQueryによってリクエストしてレンダリングする場合、
        // getServerSidePropsからキャッシュが渡っている必要がある。(でないと描画されない)
        // そのため、SSR時は常にcacheから取得されるべきで、実装意図を明確にするためPolicyも設定しておく
        // (エラーになるわけではない)
        fetchPolicy: isSSR() ? "cache-only" : "cache-first",
      },
    },
  });
}

let apolloClient: ReturnType<typeof createApolloClient>;

export function initializeApollo(
  initialState: CacheState = null
): ApolloClient<CacheState> {
  const _apolloClient = apolloClient ?? createApolloClient();

  // If your page has Next.js data fetching methods that use Apollo Client, the initial state
  // gets hydrated here
  if (initialState) {
    // Get existing cache, loaded during client side data fetching
    const existingCache = _apolloClient.extract();

    // Merge the existing cache into data passed from getStaticProps/getServerSideProps
    const data = mergeDeep(initialState, existingCache);

    // Restore the cache with the merged data
    _apolloClient.cache.restore(data);
  }
  // For SSG and SSR always create a new Apollo Client
  if (isSSR()) return _apolloClient;
  // Create the Apollo Client once in the client
  if (!apolloClient) apolloClient = _apolloClient;

  return _apolloClient;
}

export function useApollo(initialState: CacheState): ApolloClient<CacheState> {
  const store = useMemo(() => initializeApollo(initialState), [initialState]);
  return store;
}
