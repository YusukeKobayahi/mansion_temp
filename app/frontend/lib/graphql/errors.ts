export interface GraphQLError {
  graphQLErrors?: {
    message: string;
    extensions?: {
      code: string;
    };
  }[];
}

export const isNotFoundError = (err: GraphQLError): boolean => {
  const graphQLErrors = err.graphQLErrors;
  if (graphQLErrors == null) return false;

  return graphQLErrors[0]?.extensions?.code === "NOT_FOUND";
};
