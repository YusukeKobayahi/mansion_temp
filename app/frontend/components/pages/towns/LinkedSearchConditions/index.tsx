import Link from "next/link";
import { MsLibraryTownsLinkedSearchConditionsFragment } from "~/graphql/generated";
import styles from "./index.module.scss";
import { event } from "~/lib/gtag";

export interface LinkedSearchConditionsProps {
  pathname: string;
  data: MsLibraryTownsLinkedSearchConditionsFragment;
}

const LinkedSearchConditions: React.FC<LinkedSearchConditionsProps> = ({
  pathname,
  data,
}: LinkedSearchConditionsProps) => {
  if (!data.linkedSearchConditions || data.linkedSearchConditions.length === 0)
    return null;

  const clickCondition = () => {
    event({
      action: "msl_area_click",
      category: "msl_ichiran",
      label: "msl_towns_searchcondition",
      value: 1,
    });
  };

  const items = data.linkedSearchConditions.map((condition, i) => {
    const href = `${pathname}?${condition.key}=${condition.value}`;
    return (
      <Link key={i} href={href} prefetch={false} passHref>
        <button className={styles.btn} onClick={clickCondition}>
          {condition.name}
        </button>
      </Link>
    );
  });

  return (
    <section className={styles.container}>
      <div className={styles.title}>
        <p className={styles.title}>おすすめの検索条件</p>
      </div>
      <ul className={styles.list}>{items}</ul>
    </section>
  );
};

export default LinkedSearchConditions;
