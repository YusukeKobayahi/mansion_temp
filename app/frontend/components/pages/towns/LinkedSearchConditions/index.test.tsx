import { render } from "@testing-library/react";
import mockData from "~/components/pages/towns/fixtures/mockData.json";
import { TownsQuery } from "~/graphql/generated";
import LinkedSearchConditions from ".";

describe("components/pages/cities/LinkedSearchConditions", () => {
  it("正常に表示される", () => {
    const mock = mockData as TownsQuery;
    const pathname = "/ms-library/towns/38774/";
    const { asFragment } = render(
      <LinkedSearchConditions data={mock.towns[0]} pathname={pathname} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
