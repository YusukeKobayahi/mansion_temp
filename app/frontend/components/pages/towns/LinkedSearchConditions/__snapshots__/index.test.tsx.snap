// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/pages/cities/LinkedSearchConditions 正常に表示される 1`] = `
<DocumentFragment>
  <section
    class="container"
  >
    <div
      class="title"
    >
      <p
        class="title"
      >
        おすすめの検索条件
      </p>
    </div>
    <ul
      class="list"
    >
      <button
        class="btn"
        href="/ms-library/towns/38774?md=0304"
      >
        3LDK
      </button>
      <button
        class="btn"
        href="/ms-library/towns/38774?gfb=20"
      >
        タワーマンション
      </button>
      <button
        class="btn"
        href="/ms-library/towns/38774?gft=5"
      >
        低層マンション
      </button>
    </ul>
  </section>
</DocumentFragment>
`;
